components/service/glean/src/main/java/mozilla/components/service/glean/storages/StringListsStorageEngine.kt

```diff
@@ -0,0 +1,82 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.annotation.SuppressLint
+import mozilla.components.service.glean.CommonMetricData
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This singleton handles the in-memory storage logic for strings. It is meant to be used by
+ * the Specific Strings API and the ping assembling objects. No validation on the stored data
+ * is performed at this point: validation must be performed by the Specific Strings API.
+ *
+ * This class contains a reference to the Android application Context. While the IDE warns
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object StringListsStorageEngine : StringListsStorageEngineImplementation()
+
+internal open class StringListsStorageEngineImplementation(
+    override val logger: Logger = Logger("glean/StringsStorageEngine")
+) : GenericScalarStorageEngine<Array<String>?>() {
+
+    companion object {
+        // Maximum length of any list
+        private const val MAX_LIST_LENGTH_VALUE = 50
+    }
+
+    @SuppressWarnings("unchecked")
+    override fun deserializeSingleMetric(value: Any?): Array<String>? {
+        return value as? Array<String>
+    }
+
+    /**
+     * Record a string in the desired stores.
+     *
+     * @param metricData object with metric settings
+     * @param value the string value to add
+     */
+    @Synchronized
+    fun add(
+        metricData: CommonMetricData,
+        value: List<String>
+    ) {
+        // Use a custom combiner to add the string to the existing list rather than overwriting
+        super.recordScalar(metricData, value.toTypedArray()) { currentValue, newValue ->
+            currentValue?.let {
+                return@let if (it.count() < MAX_LIST_LENGTH_VALUE) {
+                    it + newValue!!
+                } else {
+                    logger.warn("List [${metricData.category}.${metricData.name}] " +
```

Comment:
nit: is "${metricData.category}.${metricData.name} - discarding new values. List limit reached." clearer? We prefer having more user-friendly messages, so that devs will not have hard time figuring out wha'ts going wrong.